### DISCUSSION TITLE
Feature Request: add read_xml() function

### DISCUSSION THREAD

**Discussion by l1t1** [duckdb/duckdb#9547] [Ideas]
Why do you want this feature?
sometime we need read xml files, such as
https://storage.googleapis.com/clickhouse_public_datasets/

**Comment by liquidcarbon** [comment]
+1
I'm running a DuckDW on 100,000s of JSONs derived from XMLs via xmltodict.  Conversion is not too difficult but every so often I wish I could query XMLs directly.

**Reply by Maxxen** [reply]
I may have something in the works...

**Reply by liquidcarbon** [reply]
tell us more? :)

**Comment by yooakim** [comment]
What is the status of the work you started [here](https://github.com/Maxxen/duckdb-xml) @Maxxen ? ;-)

**Reply by Maxxen** [reply]
Definitely something I want to pick up again, but realistically its going to be after DuckDB 1.0 :)

**Comment by david542542** [comment]
Also I believe related to this one: https://github.com/duckdb/duckdb/discussions/12218

**Comment by ansuff** [comment]
Any news on this one?

**Reply by Maxxen** [reply]
Hoping to find time to pick it up again after next release is out but I dont have a more precise timeline than that.

**Comment by Thinket** [comment]
Registering my interest in this idea! It would be very useful for ingesting data from XML-centric brokers.

**Comment by AliHmaou** [comment]
Just had to go back to Talend to load 300000 xml files (GPX). Any clue of a duckdb native solution ?

**Reply by Tishj** [reply]
https://github.com/duckdb/duckdb/discussions/9547#discussioncomment-10455241

**Reply by onnimonni** [reply]
You can already load gpx files Using the st_read() from spatial extension.

**Comment by liquidcarbon** [comment]
Thinking out loud... is this an alternative worth considering?
It's quite easy to use something like xmltodict to read into a python dict, and make a list of those.

Why can't list of dicts be registered as DuckDB tables like dataframes (which essentially _are_ lists of dicts)?
Is there a feature open for that?

**Reply by AliHmaou** [reply]
Of course it's possible to convert before handling data in duckdb but :
- it needs a "coding environment", in my case I want to stick with duckdb's SQL only spirit,
- when it comes to thousands or millions of files, the duckdb magic has no competitors in terms of performance, 
- just imagine doing "select * from xml_read(a whole directory with a bunch of xmls)" 😍

That would help many professional outcomes that are not yet "jsonified" :
- mobility : by handling KML, GPX,....
- transportation: netex
- roads : datex
- food and fisheries : un-Flux (cefact)
...

**Comment by JukkaSartinen** [comment]
This would be a good feature, almost a no-brainer.

**Comment by AliHmaou** [comment]
Guys !!! It's in the roadmap !!! https://duckdb.org/docs/stable/dev/roadmap.html

**Reply by frmadiot** [reply]
Fine, It will simplify my work !!!

**Comment by teaguesterling** [comment]
I've been working on an extension for this in the meantime: https://github.com/teaguesterling/duckdb_xml

**Reply by vhiairrassary** [reply]
Thank you for having a look, I just tested it locally 🥳. Do you plan to support HTML in addition to XML? I need to extract some data from HTML files using XPath, and I was able to test it locally with the following quick and dirty change:
```diff
diff --git a/src/xml_utils.cpp b/src/xml_utils.cpp
index 099733a..53a11ff 100644
--- a/src/xml_utils.cpp
+++ b/src/xml_utils.cpp
@@ -3,6 +3,7 @@
 #include "duckdb/common/exception.hpp"
 #include <libxml/xmlerror.h>
 #include <libxml/xmlschemas.h>
+#include <libxml/HTMLparser.h>
 #include <iostream>
 #include <functional>
 #include <set>
@@ -36,7 +37,7 @@ XMLDocRAII::XMLDocRAII(const std::string& xml_str) {
        xmlParserCtxtPtr parser_ctx = xmlNewParserCtxt();
        if (parser_ctx) {
                // Parse with default options (preserves comments and CDATA by default)
-               doc = xmlCtxtReadMemory(parser_ctx, xml_str.c_str(), xml_str.length(), nullptr, nullptr, 0);
+               doc = htmlReadMemory(xml_str.c_str(), xml_str.length(), nullptr, nullptr, HTML_PARSE_RECOVER | HTML_PARSE_NOERROR | HTML_PARSE_NOWARNING);
                xmlFreeParserCtxt(parser_ctx);
        }

```

**Reply by teaguesterling** [reply]
Would love to add this! I'll add it in before it get goes for PR into the community extensions. Thank you for the suggestion!

**Reply by teaguesterling** [reply]
Thanks again for the suggestion! I've added it in to the latest release, which I think is ready to go into a PR!

**Reply by teaguesterling** [reply]
The PR for this (along with HTML reading and utility functions) is in: https://github.com/duckdb/community-extensions/pull/496

**Reply by teaguesterling** [reply]
We renamed the extension to avoid conflicts with future work. It's now "webbed", but it's been merged into the community extensions! Thanks again for the suggestions on HTML support!

**Reply by AliHmaou** [reply]
Thanks for your work!  
Does that mean it will eventually be available through a simple "INSTALL webbed FROM community" ?  

For now, is the only option to clone the repo and compile the extension into the right directory ? 
(I haven’t had enough time to test it yet since it seems a bit complicated)

EDIT : tried it through "install webbed from community" and worked well but not on windows since it's an excluded plateform in the extension yaml. Works well on simple samples, seems to need a bit of custom definitions to extract from complex structures (seen that part of your doc but didn't try it yet). 

Well done and thank you again ! I'll keep trying and give you next feeback in your repo

**Reply by teaguesterling** [reply]
You should be able to do `INSTALL webbed FROM community`!

```
DuckDB v1.3.2 (Ossivalis) 0b83e5d2f6
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D INSTALL webbed FROM community;
D LOAD webbed;
D SELECT parse_html('<h1>Hello!</h1>');
┌───────────────────────────────────────────┐
│       parse_html('<h1>Hello!</h1>')       │
│                   html                    │
├───────────────────────────────────────────┤
│ <html><body><h1>Hello!</h1></body></html> │
└───────────────────────────────────────────┘
```

**Comment by zaneselvans** [comment]
This would be a great feature! We regularly parse hundreds of thousands of documents in an XML-based business reporting format called XBRL [using the lxml package in our FERC XBRL Extractor](https://github.com/catalyst-cooperative/ferc-xbrl-extractor) and it is painfully slow.

**Comment by teaguesterling** [comment]
It'd be great to know if the webbed community extension works for you! That
said, it's built on top of lxml but can utilize a lot of duckdb's
downstream optimizations.

**Reply by zaneselvans** [reply]
I don't know @teaguesterling but we'll take a look!

**Comment by teaguesterling** [comment]
Feel free to raise issues if you run into them. Trying to keep it useful!
